include/envoy/event/scaled_range_timer_manager.h

```diff
@@ -0,0 +1,99 @@
+  // - ScaledMinimum value:
+  //     the return value is the scale factor applied to the provided maximum.
+  // - AbsoluteMinimum:
+  //     the return value is that minimum, and the provided maximum is ignored.
+  std::chrono::milliseconds computeMinimum(std::chrono::milliseconds maximum) const {
+    struct Visitor {
+      explicit Visitor(std::chrono::milliseconds value) : value_(value) {}
+      std::chrono::milliseconds operator()(ScaledMinimum scale_factor) {
+        return std::chrono::duration_cast<std::chrono::milliseconds>(scale_factor.scale_factor_ *
+                                                                     value_);
+      }
+      std::chrono::milliseconds operator()(AbsoluteMinimum absolute_value) {
+        return absolute_value.value_;
+      }
+      const std::chrono::milliseconds value_;
+    };
+    return absl::visit<Visitor, const absl::variant<ScaledMinimum, AbsoluteMinimum>&>(
+        Visitor(maximum), *this);
+  }
+};
+
+/**
+ * Class for creating RangeTimer objects that can be adjusted towards either the minimum or maximum
+ * of their range by the owner of the manager object. Users of this class can call createTimer() to
+ * receive a new RangeTimer object that they can then enable or disable at will (but only on the
+ * same dispatcher), and setScaleFactor() to change the scaling factor. The current scale factor is
+ * applied to all timers, including those that are created later.
+ */
+class ScaledRangeTimerManager {
+public:
+  virtual ~ScaledRangeTimerManager() = default;
+
+  /**
+   * Creates a new range timer backed by the manager. The returned timer will be subject to the
+   * current and future scale factor values set on the manager. All returned timers must be deleted
+   * before the manager.
+   */
+  virtual RangeTimerPtr createRangeTimer(TimerCb callback) PURE;
+
+  /**
+   * Creates a new range timer backed by the manager that mimics the Timer interface. Calling
+   * enableTimer on the returned object sets the maximum duration, while the first argument here
+   * controls the minimum. Passing a value of ScaleFactor(x) sets the min to (x * max) when the
+   * timer is enabled, while AbsoluteValue(y) sets the min to the duration y.
+   */
+  virtual TimerPtr createTimer(ScaledTimerMinimum minimum, TimerCb callback) PURE;
+
+  /**
+   * Sets the scale factor for all timers created through this manager. The value should be between
+   * 0 and 1, inclusive. The scale factor affects the amount of time timers spend in their target
+   * range. The RangeTimers returned by createTimer will fire after (min + (max - min) *
+   * scale_factor). This means that a scale factor of 0 causes timers to fire immediately at the min
+   * duration, a factor of 0.5 causes firing halfway between min and max, and a factor of 1 causes
+   * firing at max.
+   */
+  virtual void setScaleFactor(double scale_factor) PURE;
```

Comment:
Should this take a ScaledMinimum (or similar) which can better ASSERT that it's between 0.0 and 1.0 and then clamp if needed?

Reply:
Between here and the OverloadAction class, I think it would be a good idea to have a single simple wrapper class that enforces 0.0 and 1.0 bounds. I'm going to do that in a future PR.